velox/functions/sparksql/String.h

```diff
@@ -583,4 +583,122 @@ struct OverlayVarbinaryFunction : public OverlayFunctionBase {
   }
 };
 
+/// translate(string, match, replace) -> varchar
+///
+///   Returns a new translated string. It translates the character in ``string``
+///   by a character in ``replace``. The character in ``replace`` is
+///   corresponding to the character in ``match``. The translation will
+///   happen when any character in ``string`` matching with the character in the
+///   ``match``.
+template <typename T>
+struct TranslateFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  FOLLY_ALWAYS_INLINE static std::function<int(const char*)> utf8CharCounter() {
+    return [](const char* input) { return utf8proc_char_length(input); };
+  }
+  FOLLY_ALWAYS_INLINE static std::function<int(const char*)>
+  asciiCharCounter() {
+    return [](const char*) { return 1; };
+  }
+
+  using MapType = folly::F14FastMap<StringView, StringView>;
+  MapType dictionary_;
+
+  bool isConstantTranslateDictionary = false;
+
+  FOLLY_ALWAYS_INLINE MapType buildDictionary(
+      const arg_type<Varchar>& match,
+      const arg_type<Varchar>& replace,
+      std::function<int(const char* input)> charCounter) {
+    MapType dictionary;
+    int i = 0;
+    int j = 0;
+    while (i < match.size()) {
+      StringView replacePiece;
+      // If match's character size is larger than replace's, the execessive
+      // characters in match will be removed from input string.
+      if (j >= replace.size()) {
+        replacePiece = StringView("");
+      } else {
+        int replaceCharCount = charCounter(replace.data() + j);
+        replacePiece = StringView(replace.data() + j, replaceCharCount);
+        j = j + replaceCharCount;
+      }
+      int matchCharCount = charCounter(match.data() + i);
+      StringView matchPiece = StringView(match.data() + i, matchCharCount);
+      // Only consider the first occurrence of a character in match.
+      if (dictionary.find(matchPiece) == dictionary.end()) {
+        dictionary.emplace(matchPiece, replacePiece);
+      }
+      i = i + matchCharCount;
+    }
+    return dictionary;
+  }
+
+  FOLLY_ALWAYS_INLINE void initialize(
+      const core::QueryConfig& /*config*/,
+      const arg_type<Varchar>* /*string*/,
+      const arg_type<Varchar>* match,
+      const arg_type<Varchar>* replace) {
+    if (match != nullptr && replace != nullptr) {
+      dictionary_ = buildDictionary(*match, *replace, utf8CharCounter());
+      isConstantTranslateDictionary = true;
+    }
+  }
+
+  FOLLY_ALWAYS_INLINE void doTranslate(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& input,
+      std::function<int(const char* input)> charCounter) {
```

Comment:
How fast is this implementation for all-ASCII input? I think it may be not very fast. I suggest to create a separate implementation for all-ASCII input and experiment with using an array of chars for dictionary.

Reply:
Hi @mbasmanova, currently, we use a unified map (with StringView or the suggested std::string), which can be created at `initialize()` for constant args, regardless of ASCII input or not. It looks there is no way to recognize ASCII input at `initialize()` and then do the suggested optimizations. So I'm assuming your suggested optimizations are applicable to non-constant path. But, according to the feedback from our customers, it is more common that constant args (`replace`/`match`) are used for this functions. Do we still need to do the optimization for non-constant path? Please correct me if I thought wrongly. Thanks!

Reply:
I agree that constant dictionary with all-ASCII inputs is the most common path. This is the one we should optimize for. You can easily check if all values in dictionary are all-ASCII, no?

Reply:
In the latest commit, I separated the implementations for all-Ascii path and the others. And a map with char type used is for all-Ascii input. In addition, building the map is moving from `initialize` to `call`/`callAscii` for constant args and a flag is introduced to mark whether it is already built for constant case. Thus, we can build the corresponding kind of map, regardless of constant args or not.
With your suggested optimization, around 50% time can be saved for all-Ascii input in a simple test (input string with 10k ascii characters, 100k iterations), compared with the previous implementation.